woocommerce/google-listings-and-ads
# Fix namespace for additional vendor packages

### Changes proposed in this Pull Request:

There have been additional reports of our vendor packages conflicting with other extensions. In this case `woocommerce-bookings` which both share usage of the `google/apiclient` package. 

Following is the list of challenges I came across, which resulted in having to rewrite a substantial amount of the `prefix-vendor-namespace.php` script.

#### Unique namespace match
The `apiclient` package uses the namespace `Google` as well as additional subname spaces. In itself the Google namespace is not unique enough to search for replacements. Which means that our current solution of replacing a namespace prefix was not going to work. To work around this I introduced the concept of matching "strict" or not. When strict is enabled it will match the full namespace, when it's disabled it will match any namespace within the package that starts with the package namespace.

For example:
Not strict and package name `guzzleclient` will match all namespaces like `GuzzleHttp\Psr7`, `GuzzleHttp\Guzzle`, etc...
Whereas strict and package name `apiclient` will only match namespaces that are defined as `Google`.

#### Additional namespaces
As mentioned in the previous point the `apiclient` package also uses additional subname spaces, so strict matching doesn't fully work. To get around this I introduced a list of `extra_namespaces` which can be configured per package which will also be replaced. This list might be a bit of work to maintain as we update packages but it should be relatively easy to add any additional ones.

#### Exclude file autoloads
In the previous version of the script we required any `autoload > files` configuration to be added in the main `composer.json` file. The reason being because the hash of the composer file wouldn't be unique enough if both of the plugins had the extensions in the vendor folder. However in the case of the `apiclient` package it loads an aliases.php file. Which is used for backward compatibility before namespaces were used. Since we don't need this backward compatibility (and we don't fully prefix everything in this fine), the preferable solution is to prevent this file from being autoloaded. I added the `exclude_autoload` parameter which can be configured per package and prevents these files from being required in the main `composer.json` file.

#### String replacements in JSON files
To replace namespaces in JSON files we were previously using a regular `str_replace` however this doesn't work very well to check if a namespace starts with a specific name. For example `GPBMetadata\\Google\\Protobuf\\` was getting matched because we prefix the namespace `Google`. To get around this I merged the prefixing code with the one for removing file autoloads. There we were parsing the JSON and going through it per package, which makes it easier to do some stricter matches (depending on strict mode).

#### Replacements in package composer.json
Previously we were also replacing the namespaces within the packages composer.json files (which wasn't accurate, see previous point). Since the script is called directly from composer the package map has already been created, so we only need to update the package map. So at the moment I left out the code which replaces the local composer.json files. If there is some other case where it's needed we can restore it but it needs to be improved by properly parsing the JSON files.

#### Counting uses and namespace replacements
Recently some additional debugging code was added which would count namespace and uses replacements. And would output a list of files where uses were replaced without a namespace replacement. Since the way we process replacements was changed, this list was no longer accurate since we could run replacements in packages multiple times. So I opted to remove this code.

Addresses part of #1760

### Detailed test instructions:
Luckily we have unit tests in place which cover pretty much all internal usages of these Google Services. So I relied largely on the unit tests to surface namespaces which had to be renamed. In addition to the unit tests I ran several of the smoke tests included in the WIKI to make sure API calls still functioned with the renamed packages.

To confirm that the plugin no longer conflicts with `woocommerce-bookings` the following test can be run:
1. Checkout this PR and run `rm -rf vendor && composer install`
1. Deactivate the `google-listings-and-ads` plugin
2. Install and activate [`woocommerce-bookings`](https://github.com/woocommerce/woocommerce-bookings/releases)
3. Activate the plugin through CLI `wp --allow-root plugin activate google-listings-and-ads`

> **Note**
Activating through the WP dashboard does a redirect so it doesn't encounter the situation where `woocommerce-bookings` code is active and then `google-listings-and-ads` is loaded later.
The problem is only reproducible when plugins aren't loaded alphabetically which is why we have to reproduce it with the WP CLI command.

### Changelog entry
* Fix - Prefix Google Service packages to prevent plugin conflicts.
